recorder/src/main/java/dk/dbc/solr/performance/recorder/OutputWriter.java

```diff
@@ -0,0 +1,197 @@
+import java.util.Iterator;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Consumer of {@link LogLine}s, that counts then and puts them on a
+ * {@link OutputStream}
+ *
+ * @author Morten Bøgeskov (mb@dbc.dk)
+ */
+public class OutputWriter implements AutoCloseable, Consumer<LogLine> {
+
+    private static final Logger log = LoggerFactory.getLogger(OutputWriter.class);
+
+    private final SortedSet<Entry> entries;
+    private final OutputStream os;
+    private final int orderBufferSize;
+    private long duration;
+    private final long limit;
+    private final BiConsumer<OutputStream, LogLine> firstLineMetadata;
+
+    private Instant origin;
+    private long lastEntryTimeOffset;
+    private Long timeFirstDelta;
+    private long count;
+    private boolean completed;
+
+    /**
+     * Construct a stream consumer
+     *
+     * @param os              Stream to put lines onto
+     * @param orderBufferSize how many lines should be buffered to mitigate
+     *                        kafka out of order lines
+     * @param duration        how many ms to run for
+     * @param limit           how many lines to acquire
+     */
+    public OutputWriter(OutputStream os, int orderBufferSize, long duration, long limit, BiConsumer<OutputStream, LogLine> firstLineMetadata) {
+        this.entries = new TreeSet<>();
+        this.lastEntryTimeOffset = 0L;
+        this.timeFirstDelta = null;
+        this.os = os;
+        this.orderBufferSize = orderBufferSize;
+        this.duration = duration;
+        this.limit = limit;
+        this.firstLineMetadata = firstLineMetadata;
+        this.count = 0;
+        this.completed = false;
+        log.debug("orderBufferSize = {}", orderBufferSize);
+        log.debug("duration = {}", duration);
+        log.debug("limit = {}", limit);
+    }
+
+    @Override
+    public void close() {
+        try {
+            // If not completed, but source was drained output from cache
+            log.debug("completed = {}", completed);
+            if (!completed)
+                entries.forEach(this::outputEntry);
+        } catch (CompletedException ex) {
+            log.debug("Reached limit during shutdown");
+        } finally {
+            try {
+                os.close();
+            } catch (IOException ex) {
+                log.error("Error closing output: {}", ex.getMessage());
+                log.debug("Error closing output: ", ex);
+            }
+        }
+    }
+
+    /**
+     * Stash a log line in the cache, when cache reaches it's limit, output
+     * oldest log line.
+     *
+     * @param logLine log line
+     */
+    @Override
+    public void accept(LogLine logLine) {
+        if (entries.isEmpty())
+            this.origin = logLine.getInstant();
+        long currentOffset = logLine.timeOffsetMS(origin);
```

Comment:
Won't this set the origin to whatever the current timestamp is every time the buffer is emptied, thus resetting the currentOffset back to 0?

Reply:
Entries can only be empty 1st time this is called. But if(origin == null) might be more readable

Reply:
yes - Didn't se that buffer is only emptied during close